Documentation for Druid local setup (#721)
gardener/etcd-druid: docs/development/getting-started-locally.md

# Etcd-Druid Local Setup

> This page aims to provide steps on how to setup Etcd-Druid locally with and without storage providers

Reviewer: ```suggestion
This page aims to provide steps on how to setup Etcd-Druid locally with and without storage providers.
```

## Setting up etcd-druid

> ### Clone the Etcd-Druid Github repo

Reviewer: ```suggestion
### Clone the etcd-druid github repo
```
Sub-headings need not follow upper case first letters for words

```sh
# clone the repo
git clone https://github.com/gardener/etcd-druid.git
# cd into etcd-druid folder
cd etcd-druid
```

---

> **Note:** To setup Etcd-Druid with backups enabled on a [LocalStack](https://github.com/localstack/localstack) provider, refer [this document](https://github.com/gardener/etcd-druid/blob/master/docs/development/getting-started-locally-localstack.md)

---

> ### Setting up the Kind Cluster

Reviewer 2: Can we maybe add a note here that we use a kind cluster because it's convenient, but any other k8s cluster would also suffice?

```sh
# Create a kind cluster
make kind-up
```

This creates a new kind cluster and stores the kubeconfig in the  `./hack/e2e-test/infrastructure/kind/kubeconfig` file.

To target this newly created cluster, set the `KUBECONFIG` environment variable to the kubeconfig file located at `./hack/e2e-test/infrastructure/kind/kubeconfig` by using the following

```sh
export KUBECONFIG=$PWD/hack/e2e-test/infrastructure/kind/kubeconfig
```

### Setting up Etcd-Druid

> Generates the Etcd CRD and deploy an etcd-druid pod into the cluster

Reviewer: ```suggestion
Generates the `Etcd` CRD and deploys an etcd-druid pod into the cluster
```
Use code block for `Etcd`

```sh
make deploy
```

> ### Applying the etcd CR

Reviewer 2: Add a small point after this title before the 2 points
The note would say that we can create Etcd clusters in 2 flavours, the differences between them, and that they should use one of the two based on their use case

> - **Without Backups enabled**

Reviewer 2: Can we have two sub-sections here? 
First would be `Prepare the Etcd CR` where we say the following
1. Where to find the Etcd CR
2. Comment out the `spec.backup.store` section 

Second section would be `Apply the Etcd CR` where we have the example `kubectl apply...` command

Reviewer: Agreed with @aaronfern 's point, with slight changes. You can choose to have a structure like this:
```
- Prepare the Etcd CR
  Where to find the etcd CR...
  - Without backups enabled:
    - comment out the spec.backup.store section
  - With backups enabled:
    - prepare secret
    - apply secret
    - populate etcd spec.backup.store appropriately
- Apply the Etcd CR
  How to apply the prepared etcd CR...
```

I think this flow seems quite simple, without any repetition of information/steps.

    To setup Etcd-druid without backups enabled, apply the Etcd yaml to the cluster

    Confirm that the `spec.backup.store` is commented out.

    ```sh
    # Apply the etcd CR yaml
    kubectl apply -f config/samples/druid_v1alpha1_etcd.yaml
    ```

> - **With Backups enabled (On Cloud Provider Object Stores)**

Reviewer 2: Similarly have sub-sections here
1. Prepare secret
2. Apply secret 
3. Prepare Etcd CR
4. Apply Etcd CR

    Create a secret for cloud provider access. Find the secret yaml templates for different cloud providers [here](https://github.com/gardener/etcd-backup-restore/tree/master/example/storage-provider-secrets).

    Replace the dummy values with the actual configurations and make sure to add a name and a namespace to the secret as intended.

    > **Note 1:** The secret should be applied in the same namespace as druid.
    >
    > **Note 2:** All the values in the data field of secret yaml should be in base64 encoded format.

    **Apply the secret**

    ```sh
    kubectl apply -f path/to/secret
    ``````

    Uncomment the `spec.backup.store` section of the druid yaml and set the keys to allow backuprestore to take backups by connecting to an object store.

    ```yaml
    # Configuration for storage provider
    store:
        secretRef:
            name: etcd-backup-secret-name
        container: object-storage-container-name
        provider: aws # options: aws,azure,gcp,openstack,alicloud,dell,openshift,local
        prefix: etcd-test
    ```

    Brief explanation of keys:

>   - `secretRef.name` is the name of the secret that was applied as mentioned above
>   - `store.container` is the object storage bucket name
>   - `store.provider`  is the bucket provider. Pick from the options mentioned in comment
>   - `store.prefix`    is the folder name that you want to use for your snapshots.

Reviewer 3: ```suggestion
    - `secretRef.name` is the name of the secret that was applied as mentioned above
    - `store.container` is the object storage bucket name
    - `store.provider`  is the bucket provider. Pick from the options mentioned in comment
    - `store.prefix`    is the folder name that you want to use for your snapshots.
```

    > **Note:** Before applying the Etcd yaml, make sure the bucket is created in the corresponding cloud provider

    Create the Etcd CR (Custom Resource) by applying the Etcd yaml to the cluster

    ```sh
    # Apply the etcd CR yaml
    kubectl apply -f config/samples/druid_v1alpha1_etcd.yaml
    ```

### Annotate the etcd CR

> The above step creates an Etcd resource, however etcd-druid won't pick it up for reconciliation without an annotation. To get etcd-druid to reconcile the etcd CR, annotate it with the following `gardener.cloud/operation:reconcile`.

Reviewer: Please also add a sentence about how to disable this whole annotation business, by setting druid CLI flag `--ignore-operation-annotation=true` and redeploying druid, so that every spec update will be acted upon by druid, without waiting for reconcile annotation.

Author: Thanks for the suggestion @shreyas-s-rao , I added a note even before the setup is started so that the reader will be aware of it from the start and can change the flag before starting the setup.

```sh

> # Annotate etcd-test CR to reconcile

Reviewer 2: ```suggestion
# Annotate etcd-test CR with the reconcile annotation
```

kubectl annotate etcd etcd-test gardener.cloud/operation="reconcile"
```

This starts creating the etcd cluster

> 

Reviewer: One section here to test the cluster would be beneficial to users, so that they know that the Etcd cluster they just deployed actually works as expected. You could add a section `Operate/Test the Etcd cluster` and add some steps like `kubectl get etcd -o=wide` and `etcdctl get/put` operations.

Author: I've made the changes, please verify if anything else is required. Or if I can make some modifications

## Cleaning the setup

```sh
# Delete the cluster
make kind-down
```

> This cleans up the cluster i.e deletes the created Etcd and all pods that got created along the way and also other resources such as statefulsets, services, PV's, PVC's, etc.

Reviewer 2: This says that the etcd cluster is cleared, but is not clear about the kind cluster being deleted. Can you please update this so that it's clear the kind cluster is deleted?

# Etcd-Druid Local Setup

This page aims to provide steps on how to setup Etcd-Druid locally with/without storage providers

> ## Prerequisites 

Reviewer 2: Don't we need some more prerequisites? 
maybe kind, since later on in this doc there is a section on starting a kind cluster?

Author: We need some prerequisites such as kind and helm but they will be auto installed if not present already when we trigger `make kind-up` & `make deploy` respectively. So I removed them from prequisities.

> ### [Optional] Installing [k9s](https://k9scli.io/) 

Reviewer 2: Please add the link to the k9s installation page https://k9scli.io/topics/install/

Also, according to the k9s installation page, the command for mac/linus is `brew install derailed/k9s/k9s`. Please check if it's interchangeable or not

[k9s](https://k9scli.io/) is an interactive terminal based UI to play with your kubernetes cluster. Makes it easy to interact with the cluster resources

On Mac OS & Linux, install using brew

```sh
# via Homebrew
brew install k9s
```

On Windows,
```sh
# via scoop
scoop install k9s
# via chocolatey
choco install k9s
```

> 

Reviewer 4: I agree with @seshachalam-yv. It would be ideal to keep the external dependencies to a minimum for any guide. Simple `kubectl` commands should suffice for anyone who is running `etcd-druid` locally for the first time.

> 

Author: > In this guide, `k9s` is not utilized on any of the steps. Given this, is it necessary to include it as a prerequisite?

Then can we move it out of prerequisite and put just optional heading for it? Or if it seems out of place then I can remove it.

> 

Author: Thanks for the suggestion. I removed it in the latest commit.

> ## Setting up Druid

Reviewer 2: ```suggestion
## Setting up etcd-druid
```

### Clone the Etcd-Druid Github repo
    
```sh
# clone the repo
git clone https://github.com/gardener/etcd-druid.git
# cd into etcd-druid folder
cd etcd-druid
```

---

>Note: To setup Etcd-Druid with backups enabled on a [LocalStack](https://github.com/localstack/localstack) provider, refer [this](https://github.com/gardener/etcd-druid/blob/master/docs/development/getting-started-locally-localstack.md)

---

### Setting up the Kind Cluster

```sh
# Create a kind cluster
make kind-up
```

This creates a new kind cluster and stores the kubeconfig in the  `./hack/e2e-test/infrastructure/kind/kubeconfig` file.

> To target this newly created cluster, set the `KUBECONFIG` environment variable to the kubeconfig file located at `./hack/e2e-test/infrastructure/kind/kubeconfig` by doing 

Reviewer 2: ```suggestion
To target this newly created cluster, set the `KUBECONFIG` environment variable to the kubeconfig file located at `./hack/e2e-test/infrastructure/kind/kubeconfig` by using the following 
```

```sh
export KUBECONFIG=$PWD/hack/e2e-test/infrastructure/kind/kubeconfig
```

### Setting up Etcd-Druid

> * 

Reviewer 3: Why `*` ?

Author: Thanks for pointing out. That was written earlier when there were multiple bullet points under the section, forgot to remove after refactoring. Updated in the latest commit.

>     Generates the Etcd CRD and deploy etcd-druid into the cluster

Reviewer 2: ```suggestion
    Generates the Etcd CRD and deploy an etcd-druid pod into the cluster
```

    ```sh
    make deploy
    ```

### Without Backups enabled

To setup Etcd-druid without backups enabled, apply the Etcd yaml to the cluster 
```sh

> # Apply the druid yaml

Reviewer 2: ```suggestion
# Apply the etcd CR yaml
```

kubectl apply -f config/samples/druid_v1alpha1_etcd.yaml
``` 

> Confirm that the `spec.backup.store` is commented out.

Reviewer 2: This confirmation message should be moved to before the `kubectl apply...` command

> This creates an Etcd resource but it won't be reconciled without annotation. To reconcile the Etcd, annotate `gardener.cloud/operation:reconcile` on the Etcd resource.

Reviewer 2: ```suggestion
This creates an Etcd resource, however etcd-druid won't pick it up for reconciliation without an annotation. To get etcd-druid to reconcile the etcd CR, annotate it with the following `gardener.cloud/operation:reconcile`.
```

```sh
# Annotate etcd-test CR to reconcile
kubectl annotate etcd etcd-test gardener.cloud/operation="reconcile"
```

> This should start creating the etcd cluster by rolling the pods one by one to ensure quorum is never lost.

Reviewer 2: Seems out of place to me
Maybe give some more context or remove this

Author: Meant to say that after the reconcilation of the etcd CR, it starts creating the etcd cluster in a rolling fashion. I will instead just write "This starts creating the etcd cluster" without specifying the rolling of pods. Is this ok?

> ### With Backups enabled on an actual cloud provider

Reviewer 2: You can combine this with the previous heading (`### Without Backups enabled`)

Have one main heading (for ex `Applying the etcd CR`) and have these 2 as sub headings

Author: Agree!

Create a secret for cloud provider access. Find the secret yaml templates for different cloud providers [here](https://github.com/gardener/etcd-backup-restore/tree/master/example/storage-provider-secrets). 

> Replace the dummy values with the actual configurations and make sure to name and namespace the secret as intended. 

Reviewer 2: ```suggestion
Replace the dummy values with the actual configurations and make sure to add a name and a namespace to the secret as intended. 
```

>Note 1): The secret should be applied in the same namespace as druid.
>
>Note 2): All the values in the data field of secret yaml should be in base64 encoded format.

<b>Apply the secret</b> `kubectl apply -f path_to_secret`

Uncomment the `spec.backup.store` section of the druid yaml and set the keys to allow backuprestore to take backups by connecting to an object store. 

```yaml
# Configuration for storage provider
store:
    secretRef:
        name: etcd-backup-secret-name
    container: object-storage-container-name
    provider: aws # options: aws,azure,gcp,openstack,alicloud,dell,openshift,local
    prefix: etcd-test
```
Brief explanation of keys:

* `secretRef.name` is the name of the secret that was applied as mentioned above
* `store.container` is the object storage bucket name
* `store.provider`  is the bucket provider. Pick from the options mentioned in comment
* `store.prefix`    is the folder name that you want to use for your snapshots.

> Note: Before applying the Etcd yaml, make sure the bucket is created in the corresponding cloud provider

Create the Etcd CR (Custom Resource) by applying the Etcd yaml to the cluster 
```sh

> # Apply the druid yaml

Reviewer 2: ```suggestion
# Apply the etcd CR yaml
```

kubectl apply -f config/samples/druid_v1alpha1_etcd.yaml
```

> Reconcile the controller to create Etcd cluster by annotating the Etcd CR with `gardener.cloud/operation:reconcile`.

Reviewer 2: Please change it to something similar to the section above

```sh
# Annotate etcd-test CR to reconcile
kubectl annotate etcd etcd-test gardener.cloud/operation="reconcile"
```

## Cleaning the setup

```sh
# Delete the cluster
make kind-down
```

> This deletes the created Etcd and all pods that got created along the way and also other resources such as statefulsets, services, PV's, PVC's, etc.

Reviewer 2: Please mention that this also cleans up the kind cluster

# Etcd-Druid Local Setup

This page aims to provide steps on how to setup Etcd-Druid locally with and without storage providers.

## Clone the etcd-druid github repo

```sh
# clone the repo
git clone https://github.com/gardener/etcd-druid.git
# cd into etcd-druid folder
cd etcd-druid
```

---

> **Note:**
>
>- Etcd-druid uses [kind](https://kind.sigs.k8s.io/) as it's local Kubernetes engine. The local setup is configured for kind due to it's convenience but any other kubernetes setup would also work.
>- To setup Etcd-Druid with backups enabled on a [LocalStack](https://github.com/localstack/localstack) provider, refer [this document](getting-started-locally-localstack.md)

> >- In the section [Annotate Etcd CR with the reconcile annotation](#annotate-etcd-cr-with-the-reconcile-annotation), the flag `ignore-operation-annotation` is set to false, which means manual annotation to the Etcd CR is requied to reconcile the Etcd CR. To disable this and auto reconcile the Etcd CR for any change in the Etcd yaml, set the`ignoreOperationAnnotation` flag to `true` in the `values.yaml` located at `$PWD/charts/druid/values.yaml`. Or if the Etcd-Druid is being run as a process, then while starting the process, attach the CLI flag `--ignore-operation-annotation=true` to it.

Reviewer: ```suggestion
>- In the section [Annotate Etcd CR with the reconcile annotation](#annotate-etcd-cr-with-the-reconcile-annotation), the flag `ignore-operation-annotation` is set to false, which means a special annotation is required  on the Etcd CR, for etcd-druid to reconcile it. To disable this behavior and allow auto-reconciliation of the Etcd CR for any change in the Etcd spec, set the `ignoreOperationAnnotation` flag to `true` in the `values.yaml` located at [`charts/druid/values.yaml`](../../charts/druid/values.yaml). Or if etcd-druid is being run as a process, then while starting the process, set the CLI flag `--ignore-operation-annotation=true` for it.
```

### Prepare the Etcd CR

Etcd CR can be configured in 2 ways. Either to take backups to the store or disable them. Follow the appropriate section below based on the requirement.

The Etcd CR can be found at this location `$PWD/config/samples/druid_v1alpha1_etcd.yaml`

- **Without Backups enabled**

    To setup Etcd-druid without backups enabled, make sure the `spec.backup.store` section of the Etcd CR is commented out.

- **With Backups enabled (On Cloud Provider Object Stores)**

  - **Prepare the secret**

    Create a secret for cloud provider access. Find the secret yaml templates for different cloud providers [here](https://github.com/gardener/etcd-backup-restore/tree/master/example/storage-provider-secrets).

    Replace the dummy values with the actual configurations and make sure to add a name and a namespace to the secret as intended.

    > **Note 1:** The secret should be applied in the same namespace as druid.
    >
    > **Note 2:** All the values in the data field of secret yaml should be in base64 encoded format.

  - **Apply the secret**

    ```sh
    kubectl apply -f path/to/secret
    ```

>   - **Update Etcd**

Reviewer: ```suggestion
  - **Adapt `Etcd` resource**
```

The term `update` can be misread as "update the etcd resource using kubectl", rather than "uncomment the section within the etcd manifest file". Hence, you can probably use `Adapt resource` as a more unambiguous word

> **Note :** If the `ignore-operation-annotation` flag is set to `true`, this step is not required

The above step creates an Etcd resource, however etcd-druid won't pick it up for reconciliation without an annotation. To get etcd-druid to reconcile the etcd CR, annotate it with the following `gardener.cloud/operation:reconcile`.

```sh
# Annotate etcd-test CR to reconcile
kubectl annotate etcd etcd-test gardener.cloud/operation="reconcile"
```

This starts creating the etcd cluster

### Verify the Etcd cluster

To obtain information regarding the newly instantiated etcd cluster, perform the following step, which gives details such as the cluster size, readiness status of its members, and various other attributes.

```sh
kubectl get etcd -o=wide
```

#### Verify Etcd Member Pods

To check the etcd member pods, do the following and look out for pods starting with the name `etcd-`

```sh
kubectl get pods
```

#### Verify Etcd Pods' Functionality

Verify the working conditions of the etcd pods by putting data through a etcd container and access the db from same/another container depending on single/multi node etcd cluster

```sh
# Put a key-value pair into the etcd 

> kubectl exec -it <etcd_pod> -c etcd -- etcdctl put key1 value1

Reviewer: We cannot assume that `exec` will work, specifically for clusters using etcd-wrapper. Instead, You can add a line before this to either exec into the etcd container with a bash session using `kubectl exec -it ... -- bash`, or start an ephemeral container and attach it to the etcd container using `kubectl debug ...` (you may also link to [this doc](https://github.com/gardener/etcd-wrapper/blob/main/docs/deployment/ops.md#using-the-image)). Once that's done, you can simply specify the `etcdctl` commands directly.

Author: Thanks for pointing it out @shreyas-s-rao. Updated in the latest commit.

# Retrieve all key-value pairs in the etcd db
kubectl exec -it <etcd_pod> -c etcd -- etcdctl get --prefix ""
```

If multinode etcd cluster, insert the key-value pair from `etcd` container of one etcd member and retrieve from `etcd` container of another to verify consensus between the multiple etcd members.

#### View Etcd Database File

To inspect the Etcd database file, execute the following command in the `backup-restore` container of any Etcd pods to check the database file located at `var/etcd/data/new.etcd/snap/db`

```sh

> kubectl exec -it <etcd_pod> -c backup-restore -- cat var/etcd/data/new.etcd/snap/db

Reviewer: Same as above. You need not specify the steps to get into the container, but simply link to the previous section for that information.

kubectl get pods
```

> #### Verify Etcd Pod's Functionality

Reviewer: You can continue using `Pods'`, since the section talks about verifying multiple pods' functionality.

Author: Yeah makes sense.
